src/main/java/net/minecraftforge/common/ForgeHooks.java

```diff
@@ -1334,17 +1339,27 @@ public static void readAdditionalLevelSaveData(CompoundTag rootTag)
                     continue;
                 }
                 String modVersion = mod.getString("ModVersion");
-                Optional<? extends ModContainer> container = ModList.get().getModContainerById(modId);
-                if (container.isEmpty())
-                {
-                    LOGGER.error(WORLDPERSISTENCE,"This world was saved with mod {} which appears to be missing, things may not work well", modId);
-                    continue;
-                }
-                if (!Objects.equals(modVersion, MavenVersionStringHelper.artifactVersionToString(container.get().getModInfo().getVersion())))
-                {
-                    LOGGER.warn(WORLDPERSISTENCE,"This world was saved with mod {} version {} and it is now at version {}, things may not work well", modId, modVersion, MavenVersionStringHelper.artifactVersionToString(container.get().getModInfo().getVersion()));
-                }
+                ModList.get().getModContainerById(modId).ifPresentOrElse(container -> {
+                    final var loadingVersion = container.getModInfo().getVersion();
+                    final var previousVersion = new DefaultArtifactVersion(modVersion);
+                    if (!loadingVersion.equals(previousVersion))
+                    {
+                        // Enqueue mismatched versions for bulk event
+                        mismatchedVersions.put(modId, previousVersion);
+                    }
+                }, () -> LOGGER.error(WORLDPERSISTENCE,"This world was saved with mod {} which appears to be missing, things may not work well", modId));
```

Comment:
Would it not make sense to store the missing mods as well. Just to give the event full context? "World had ModX at V, but its now missing"